snuba/cli/optimize.py

```diff
@@ -71,5 +80,21 @@ def optimize(
             ClickhouseClientSettings.OPTIMIZE
         )
 
-    num_dropped = run_optimize(connection, storage, database, before=today)
+    if not ignore_cutoff:
+        # Adding 10 minutes to the current time before finding the midnight time
+        # to ensure this keeps working even if the system clock of the host that
+        # starts the pod is slightly ahead of the system clock of the host running
+        # the job. This prevents us from getting the wrong midnight.
+        last_midnight = (datetime.now() + timedelta(minutes=10)).replace(
+            hour=0, minute=0, second=0, microsecond=0
+        )
+        cutoff: Optional[datetime] = last_midnight + settings.OPTIMIZE_JOB_CUTOFF_TIME
```

Comment:
is there a need to plumb this through the whole stack? if we're doing the work of calculating what the cuttoff is here, why not enforce it here as well?

Reply:
we cannot. 
The goal is not to prevent the entire job from starting past 11 pm. The goal is to ensure that a running job stops by 11 pm in case it is that time and it still has work to do. 
That can only be enforced in the `optimize_partitions` call, because there is when we execute the `OPTIMIZE` statement for each partition one by one. That function takes hours to execute as it runs each OPTIMIZE. So it is the place where to pending work if it is 11 pm